File: labs/control-ui/api-shadow/dns_tunnel_codec.ts
export interface DnsTunnelFrame {
  sessionId: number;
  sequence: number;
  isFinal: boolean;
  payload: Uint8Array;
}

export class DnsTunnelCodec {
  public static encodeToQuery(frame: DnsTunnelFrame, domainSuffix: string): string {
    const raw = new Uint8Array(5 + frame.payload.length);
    raw[0] = (frame.sessionId >> 8) & 0xff;
    raw[1] = frame.sessionId & 0xff;
    raw[2] = (frame.sequence >> 8) & 0xff;
    raw[3] = frame.sequence & 0xff;
    raw[4] = frame.isFinal ? 1 : 0;
    raw.set(frame.payload, 5);

    const hex = Array.from(raw).map(b => b.toString(16).padStart(2, '0')).join('');
    return `${hex}.${domainSuffix}`;
  }

  public static decodeFromQuery(query: string, domainSuffix: string): DnsTunnelFrame | null {
    if (!query.endsWith(domainSuffix)) return null;
    let trimmed = query.slice(0, query.length - domainSuffix.length);
    if (trimmed.endsWith('.')) trimmed = trimmed.slice(0, -1);
    const label = trimmed.split('.')[0] ?? '';
    if (label.length < 10 || label.length % 2 !== 0) return null;

    const raw = new Uint8Array(label.length / 2);
    for (let i = 0; i < raw.length; i++) {
      raw[i] = parseInt(label.slice(i * 2, i * 2 + 2), 16);
    }

    const sessionId = ((raw[0] ?? 0) << 8) | (raw[1] ?? 0);
    const sequence = ((raw[2] ?? 0) << 8) | (raw[3] ?? 0);
    const isFinal = (raw[4] ?? 0) === 1;
    const payload = raw.slice(5);

    return { sessionId, sequence, isFinal, payload };
  }
}

File: labs/mobile/donor-archive/tunnel/DnsTunnelCodec.kt
// Copyright 2024-2026 LumiNet Authors. All Rights Reserved.
package com.luminet.android.tunnel

import java.nio.ByteBuffer

data class DnsTunnelFrame(
    val sessionId: Int,
    val sequence: Int,
    val isFinal: Boolean,
    val payload: ByteArray
)

object DnsTunnelCodec {
    fun encodeToQuery(frame: DnsTunnelFrame, domainSuffix: String): String {
        val buf = ByteBuffer.allocate(5 + frame.payload.size)
        buf.putShort(frame.sessionId.toShort())
        buf.putShort(frame.sequence.toShort())
        buf.put(if (frame.isFinal) 1.toByte() else 0.toByte())
        buf.put(frame.payload)

        val hex = buf.array().joinToString("") { "%02x".format(it) }
        return "$hex.$domainSuffix"
    }

    fun decodeFromQuery(query: String, domainSuffix: String): DnsTunnelFrame? {
        if (!query.endsWith(domainSuffix)) return null
        val trimmed = query.removeSuffix(domainSuffix).removeSuffix(".")
        val label = trimmed.split(".").firstOrNull() ?: return null
        if (label.length < 10 || label.length % 2 != 0) return null

        val bytes = ByteArray(label.length / 2)
        for (i in bytes.indices) {
            bytes[i] = label.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }

        val buf = ByteBuffer.wrap(bytes)
        val sess = buf.short.toInt() and 0xFFFF
        val seq = buf.short.toInt() and 0xFFFF
        val isFinal = buf.get() == 1.toByte()
        val payload = ByteArray(buf.remaining())
        buf.get(payload)

        return DnsTunnelFrame(sess, seq, isFinal, payload)
    }
}
